### PR TITLE
DDF-1881 Microsoft input transformer.

### PR DESCRIPTION
**Pull request by shane-voisard**
#### What does this PR do?
Request review of initial design and architecture.

#### Who is reviewing it?
@jaymcnallie @kcwire @bdeining @jlcsmith  @glenhein

#### How should this be tested?
Not ready for testing yet.

#### Any background context you want to provide?
- build completes with -Djacoco.skip=true; i.e. unit testing needs work.
- does not include ingested doc into metacard metadata <body> tag.
- uses Clavin to parse ingested doc for geo location.
- Clavin encapsulated into a bundle.
- Clavin requires specific/different version of lucene index.
- ClavinUpdateCommand supports admin creation of dedicated clavin lucene index.

#### What are the relevant tickets?
https://codice.atlassian.net/browse/DDF-1881

#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

### CODE REVIEW

File: catalog/catalog-app/src/main/resources/features.xml [hunk 1]
```diff
@@ -211,6 +211,11 @@
         <bundle>mvn:ddf.catalog.transformer/video-input-transformer/${project.version}</bundle>
     </feature>
 
+    <feature name="catalog-transformer-microsoft" install="manual" version="${project.version}"
+             description="Microsoft Word Input Transformer.">
```
Comment (glenhein) [inline catalog/catalog-app/src/main/resources/features.xml]: The name of this input transformer makes it sound like it is for all microsoft file types, but the description restricts it to Word docs. Should the name be catalog-transformer-msword?

File: catalog/spatial/geocoding/spatial-clavin-lucene/src/main/java/org/codice/ddf/spatial/clavin/ClavinFacade.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+/**
+ * A {@code ClavinFacade} provides access to the clavin lucene index.
+ */
+public interface ClavinFacade {
+    /**
+     * Get a list of clavin ResolvedLocation.
+     *
+     * @throws ClavinFacadeException if an error occurs while accessing the index.
+     */
+    List<ResolvedLocation> getResolvedLocations(String document) throws ClavinFacadeException;
```
Comment (glenhein) [inline catalog/spatial/geocoding/spatial-clavin-lucene/src/main/java/org/codice/ddf/spatial/clavin/ClavinFacade.java]: Is "document" a required parameter? Please document what the parameter means, and what happens if null is passed in.

File: catalog/spatial/geocoding/spatial-clavin-lucene/src/main/java/org/codice/ddf/spatial/clavin/ClavinFacade.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+    /**
+     * Creates clavin lucene index from resource.
+     *
+     * @throws ClavinFacadeException if an error occurs while accessing the index.
+     */
+    void createIndex(String resource) throws ClavinFacadeException;
```
Comment (glenhein) [inline catalog/spatial/geocoding/spatial-clavin-lucene/src/main/java/org/codice/ddf/spatial/clavin/ClavinFacade.java]: Is "resource" a required parameter? Please document what the parameter means, and what happens if null is passed in.

File: catalog/spatial/geocoding/spatial-clavin-lucene/src/main/java/org/codice/ddf/spatial/clavin/ClavinFacadeImpl.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+import com.bericotech.clavin.resolver.ResolvedLocation;
+
+public class ClavinFacadeImpl implements ClavinFacade {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ClavinFacadeImpl.class);
+
+    private String clavinIndexLocation;
+
+    private GeoParser clavinParser;
+
+    public void setClavinIndexLocation(final String indexLocation) {
```
Comment (glenhein) [inline catalog/spatial/geocoding/spatial-clavin-lucene/src/main/java/org/codice/ddf/spatial/clavin/ClavinFacadeImpl.java]: Since this a public method, please add some javadoc, and test indexLocation for null values (if appropriate).

File: catalog/spatial/geocoding/spatial-clavin-lucene/src/main/java/org/codice/ddf/spatial/clavin/ClavinFacadeImpl.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        this.clavinIndexLocation = indexLocation;
+    }
+
+    @Override
+    public List<ResolvedLocation> getResolvedLocations(String document)
```
Comment (glenhein) [inline catalog/spatial/geocoding/spatial-clavin-lucene/src/main/java/org/codice/ddf/spatial/clavin/ClavinFacadeImpl.java]: If the contract in the interface is modified to declare the nullibility of "document", then add the appropriate code to test for the null status.

File: catalog/spatial/geocoding/spatial-clavin-lucene/src/main/java/org/codice/ddf/spatial/clavin/ClavinFacadeImpl.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        try {
+            return clavinParser.parse(document);
+        } catch (Exception e) {
+            // clavin Geoparser#parse throws Exception !
+            throw new ClavinFacadeException("Clavin failed to resolve geo location. ", e);
+        }
+    }
+
+    @Override
+    public void createIndex(String resource) throws ClavinFacadeException {
```
Comment (glenhein) [inline catalog/spatial/geocoding/spatial-clavin-lucene/src/main/java/org/codice/ddf/spatial/clavin/ClavinFacadeImpl.java]: If the contract in the interface is modified to declare the nullibility of "resource", then add the appropriate code to test for the null status.

File: catalog/spatial/geocoding/spatial-clavin-lucene/src/test/java/org/codice/ddf/spatial/clavin/TestClavinFacade.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        ClavinFacadeImpl cli = new ClavinFacadeImpl();
+
+        cli.setClavinIndexLocation(INDEX_PATH);
+
+        cli.createIndex(RESOURCE_PATH + "IO.txt");
+
+        List<ResolvedLocation> resolvedLocationList = cli.getResolvedLocations(document);
+
+        //assertTrue(!resolvedLocationList.isEmpty());
+        assertTrue(true);
```
Comment (glenhein) [inline catalog/spatial/geocoding/spatial-clavin-lucene/src/test/java/org/codice/ddf/spatial/clavin/TestClavinFacade.java]: why call assertTrue(true)?

File: catalog/spatial/spatial-commands/src/main/java/org/codice/ddf/commands/spatial/geonames/ClavinUpdateCommand.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+            required = true)
+    private String resource = null;
+
+    @Option(name = "-c", aliases = "--create",
+            description = "Create a new index, overwriting any existing index at the destination.")
+    private boolean create;
+
+    private ClavinFacade clavinFacade;
+
+    public void setClavinFacade(final ClavinFacade clavinFacade) {
```
Comment (glenhein) [inline catalog/spatial/spatial-commands/src/main/java/org/codice/ddf/commands/spatial/geonames/ClavinUpdateCommand.java]: Should there be some error checking on the clavinFacade parameter?

File: catalog/spatial/spatial-commands/src/main/java/org/codice/ddf/commands/spatial/geonames/ClavinUpdateCommand.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+        this.clavinFacade = clavinFacade;
+    }
+
+    public void setResource(String resource) {
```
Comment (glenhein) [inline catalog/spatial/spatial-commands/src/main/java/org/codice/ddf/commands/spatial/geonames/ClavinUpdateCommand.java]: Should there be some error checking on the resource parameter?

File: catalog/transformer/catalog-transformer-microsoft/pom.xml [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>ddf.catalog.transformer</groupId>
+        <artifactId>transformer</artifactId>
+        <version>2.9.0-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>microsoft-input-transformer</artifactId>
```
Comment (glenhein) [inline catalog/transformer/catalog-transformer-microsoft/pom.xml]: It would be nice to see an artifactId that more is more accurate. eg. input-transformer-msword

File: catalog/transformer/catalog-transformer-microsoft/src/main/java/ddf/catalog/transformer/input/MicrosoftTransformer.java [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+import ddf.catalog.data.Metacard;
+import ddf.catalog.data.impl.AttributeImpl;
+import ddf.catalog.transform.CatalogTransformerException;
+import ddf.catalog.transform.InputTransformer;
+import ddf.catalog.transformer.common.tika.MetacardCreator;
+import ddf.catalog.transformer.common.tika.TikaMetadataExtractor;
+
+import net.sf.saxon.TransformerFactoryImpl;
+
+public class MicrosoftTransformer implements InputTransformer {
```
Comment (glenhein) [inline catalog/transformer/catalog-transformer-microsoft/src/main/java/ddf/catalog/transformer/input/MicrosoftTransformer.java]: I would prefer a class name like MswordInputTransformer.

File: catalog/transformer/catalog-transformer-microsoft/src/main/java/ddf/catalog/transformer/input/MicrosoftTransformer.java [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+    private static final Logger LOGGER = LoggerFactory.getLogger(MicrosoftTransformer.class);
+
+    private Templates templates = null;
+
+    private ClavinFacade clavinFacade;
+
+    public MicrosoftTransformer(BundleContext bundleContext, ClavinFacade clavinFacade) {
```
Comment (glenhein) [inline catalog/transformer/catalog-transformer-microsoft/src/main/java/ddf/catalog/transformer/input/MicrosoftTransformer.java]: Add some javadoc on the ctor. Can the parameters be null? What happens if they are null?

File: catalog/transformer/catalog-transformer-microsoft/src/main/java/ddf/catalog/transformer/input/MicrosoftTransformer.java [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+        this.clavinFacade = clavinFacade;
+
+        try {
+            templates = new TransformerFactoryImpl().newTemplates(new StreamSource(
```
Comment (glenhein) [inline catalog/transformer/catalog-transformer-microsoft/src/main/java/ddf/catalog/transformer/input/MicrosoftTransformer.java]: Since your code does a lot of the same logic as TikaInputTransformer, can you refactor both classes to pull out the common code into a utility/support class?

File: catalog/transformer/catalog-transformer-microsoft/src/main/java/ddf/catalog/transformer/input/MicrosoftTransformer.java [hunk 14]
```diff
@@ -0,0 +1,296 @@
...
+    @Override
+    public Metacard transform(InputStream input, String id)
+            throws IOException, CatalogTransformerException {
+        LOGGER.debug("Transforming input stream using Tika.");
+
+        if (input == null) {
+            throw new CatalogTransformerException("Cannot transform null input.");
+        }
+
+        try (FileBackedOutputStream fileBackedOutputStream = new FileBackedOutputStream(1000000)) {
```
Comment (glenhein) [inline catalog/transformer/catalog-transformer-microsoft/src/main/java/ddf/catalog/transformer/input/MicrosoftTransformer.java]: I'd like to see 1000000 made a static final constant.

File: catalog/transformer/catalog-transformer-microsoft/src/main/java/ddf/catalog/transformer/input/MicrosoftTransformer.java [hunk 15]
```diff
@@ -0,0 +1,296 @@
...
+
+    private void createThumbnail(InputStream input, Metacard metacard) {
+        try {
+            Image image = ImageIO.read(new CloseShieldInputStream(input));
+
+            if (null != image) {
+                BufferedImage bufferedImage = new BufferedImage(image.getWidth(null),
+                        image.getHeight(null), BufferedImage.TYPE_INT_RGB);
+                Graphics2D graphics = bufferedImage.createGraphics();
+                graphics.drawImage(image, null, null);
```
Comment (glenhein) [inline catalog/transformer/catalog-transformer-microsoft/src/main/java/ddf/catalog/transformer/input/MicrosoftTransformer.java]: Should you put the graphics.dispose() call in a finally block?

File: catalog/transformer/catalog-transformer-microsoft/src/main/java/ddf/catalog/transformer/input/MicrosoftTransformer.java [hunk 16]
```diff
@@ -0,0 +1,296 @@
...
+                throws SAXException {
+            if (!this.insideBodyTag()) {
+                super.startElement(uri, localName, qName, atts);
+            }
+            /*
+                - once found, stop checking.
+                - check after, so empty body tag is output
+             */
+            if (!bodyStartTag) {
+                bodyStartTag = "body".equals(localName);
```
Comment (glenhein) [inline catalog/transformer/catalog-transformer-microsoft/src/main/java/ddf/catalog/transformer/input/MicrosoftTransformer.java]: Should this string comparison be case-insensitive?

File: catalog/transformer/catalog-transformer-microsoft/src/main/java/ddf/catalog/transformer/input/MicrosoftTransformer.java [hunk 17]
```diff
@@ -0,0 +1,296 @@
...
+        }
+
+        @Override
+        public void endElement(String uri, String localName, String qName) throws SAXException {
+            /*
+                - once found, stop checking.
+                - check before, so empty body tag is output
+             */
+            if (!bodyEndTag) {
+                bodyEndTag = "body".equals(localName);
```
Comment (glenhein) [inline catalog/transformer/catalog-transformer-microsoft/src/main/java/ddf/catalog/transformer/input/MicrosoftTransformer.java]: Should this string comparison be case-insensitive?

Also, since "body" is used twice, you may want to make it a static final constant.

File: catalog/transformer/catalog-transformer-microsoft/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+            <entry key="id" value="msword"/>
+            <!-- shortname only exists for backwards compatibility -->
+            <entry key="shortname" value="msword"/>
+            <entry key="description"
+                   value="Detects and extracts metadata microsoft word documents."/>
+            <entry key="mime-type">
+                <list>
+                    <value>application/msword</value>
+                    <value>application/vnd.ms-excel</value>
+                    <value>application/vnd.ms-powerpoint</value>
```
Comment (glenhein) [inline catalog/transformer/catalog-transformer-microsoft/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: Are Excel and Powerpoint actually MS Word documents? Or are they MS Office documents? Does your code only work with old-style (pre openoffice xml) documents?

File: catalog/transformer/catalog-transformer-microsoft/src/main/java/ddf/catalog/transformer/input/MicrosoftTransformer.java (hunk 12 above)
Comment (glenhein) [inline catalog/transformer/catalog-transformer-microsoft/src/main/java/ddf/catalog/transformer/input/MicrosoftTransformer.java]: How does your code handle password protected documents?

**Comment by glenhein** [comment]
Just some thoughts ... you have essentially two different things going on here. First, you have a general purpose input transformer based on tika that strips out the body tag. Second, you have an input transformer just for ms docs that adds in geo data. You could create two separate input transformers. First, BetterTikaInputTransformer and MsWordInputTransformer. The MsWordInputTransformer would have a ctor that takes another input transformer (decorator pattern). If the geo stuff is also general, and not unique to just ms word docs, then you could make a third transformer, GeoStuffInputTransformer, which would also be a decorator. This is basically what I'm doing with my PptxInputTransformer.

File: catalog/transformer/catalog-transformer-microsoft/src/main/java/ddf/catalog/transformer/input/MicrosoftTransformer.java [hunk 19]
```diff
@@ -0,0 +1,296 @@
...
+            String metadataText = handler.toString();
+            if (templates != null) {
+                metadataText = transformToXml(metadataText);
+            }
+
+            Metacard metacard = MetacardCreator.createBasicMetacard(metadata, id, metadataText);
+
+            try (InputStream is = fileBackedOutputStream.asByteSource().openStream()) {
+                try {
+                    String document = IOUtils.toString(is, "UTF-8");
```
Comment (millerw8) [inline catalog/transformer/catalog-transformer-microsoft/src/main/java/ddf/catalog/transformer/input/MicrosoftTransformer.java]: Why bother with a filebackedOutputStream if you are going to pull it all into memory anyway?

File: catalog/spatial/geocoding/spatial-clavin-lucene/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+<blueprint xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+           xmlns="http://www.osgi.org/xmlns/blueprint/v1.0.0"
+           xsi:schemaLocation="http://www.osgi.org/xmlns/blueprint/v1.0.0
+               http://www.osgi.org/xmlns/blueprint/v1.0.0/blueprint.xsd"
+           xmlns:cm="http://aries.apache.org/blueprint/xmlns/blueprint-cm/v1.1.0">
+
+    <bean id="clavinFacade" class="org.codice.ddf.spatial.clavin.ClavinFacadeImpl">
+        <cm:managed-properties persistent-id="org.codice.ddf.spatial.clavin.index.properties"
+                               update-strategy="container-managed" />
+        <argument ref="clavinFacade"/>
```
Comment (millerw8) [inline catalog/spatial/geocoding/spatial-clavin-lucene/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: What is the argument for?

File: catalog/spatial/geocoding/spatial-clavin-lucene/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ *
+ **/ -->
+<blueprint xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+           xmlns="http://www.osgi.org/xmlns/blueprint/v1.0.0"
+           xsi:schemaLocation="http://www.osgi.org/xmlns/blueprint/v1.0.0
+               http://www.osgi.org/xmlns/blueprint/v1.0.0/blueprint.xsd"
+           xmlns:cm="http://aries.apache.org/blueprint/xmlns/blueprint-cm/v1.1.0">
+
+    <bean id="clavinFacade" class="org.codice.ddf.spatial.clavin.ClavinFacadeImpl">
```
Comment (millerw8) [inline catalog/spatial/geocoding/spatial-clavin-lucene/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: I assume the life cycle for this is long right?  The index only gets read once?

**Comment by millerw8** [comment]
I agree with @glenhein's observations, and would like to go a step further.  I feel like the addition of Clavin should not be Microsoft related at all.  At the end of the day, we are trying to extract metadata from unstructured text.  I would like to see a service interface that offers that capability, and all transformers (ppt/pptx, pdf, doc/docx, txt, etc...) be able to grab all of these services and invoke them.

What does Clavin do really?  Recognize a geoname and provide a point associated with it?  Not really that powerful.

What happens when we want named entity resolution as well.  Or recognize lat/lngs specifically mentioned in the document, or recognize dates in the document.  It would be nice to have additional enrichment services available that could process on all unstructured documents.

File: catalog/transformer/catalog-transformer-microsoft/src/main/java/ddf/catalog/transformer/input/MicrosoftTransformer.java (hunk 19 above)
Comment (stustison) [inline catalog/transformer/catalog-transformer-microsoft/src/main/java/ddf/catalog/transformer/input/MicrosoftTransformer.java]: This can't be in here and will blow up the heap. Do not use IOUtils.toString or IOUtils.toByteArray

File: catalog/spatial/geocoding/spatial-clavin-lucene/src/main/java/org/codice/ddf/spatial/clavin/ClavinFacadeImpl.java (hunk 6 above)
Comment (millerw8) [inline catalog/spatial/geocoding/spatial-clavin-lucene/src/main/java/org/codice/ddf/spatial/clavin/ClavinFacadeImpl.java]: Does this ever get called when the system is restarted?  It seems you only create the calvinParser when the command line interface selects a file to index.

File: catalog/spatial/geocoding/pom.xml [hunk 22]
```diff
@@ -33,5 +33,6 @@
         <module>spatial-geocoding-localsearch</module>
         <module>spatial-geocoding-websearch</module>
         <module>spatial-geocoding-admin-module</module>
+        <module>spatial-clavin-lucene</module>
```
Comment (bdeining) [inline catalog/spatial/geocoding/pom.xml]: `spatial-geocoding-clavin`?  I feel like we should keep the naming convention the same in the geocoding module.  Maybe this would fit into the `spatial-geocoding-lucene` package?

File: catalog/spatial/geocoding/spatial-clavin-lucene/pom.xml [hunk 23]
```diff
@@ -0,0 +1,170 @@
...
+                        <configuration>
+                            <haltOnFailure>true</haltOnFailure>
+                            <rules>
+                                <rule>
+                                    <element>BUNDLE</element>
+                                    <limits>
+                                        <limit>
+                                            <counter>INSTRUCTION</counter>
+                                            <value>COVEREDRATIO</value>
+                                            <!-- <minimum>0.70</minimum> -->
```
Comment (bdeining) [inline catalog/spatial/geocoding/spatial-clavin-lucene/pom.xml]: Comments can be removed

File: catalog/spatial/geocoding/spatial-clavin-lucene/src/test/java/org/codice/ddf/spatial/clavin/TestClavinFacade.java [hunk 24]
```diff
@@ -0,0 +1,80 @@
...
+        assertFalse("The 'clavin/index' directory should not exist.",
+                Files.exists(Paths.get(INDEX_PATH)));
+
+        ClavinFacadeImpl cli = new ClavinFacadeImpl();
+
+        cli.setClavinIndexLocation(INDEX_PATH);
+
+        cli.createIndex(RESOURCE_PATH + "IO.txt");
+
+        assertTrue(Files.exists(Paths.get(INDEX_PATH)));
```
Comment (bdeining) [inline catalog/spatial/geocoding/spatial-clavin-lucene/src/test/java/org/codice/ddf/spatial/clavin/TestClavinFacade.java]: Recommend using assertThat vs assertTrue

File: catalog/spatial/geocoding/spatial-clavin-lucene/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 25]
```diff
@@ -0,0 +1,49 @@
...
+        <cm:managed-properties persistent-id="org.codice.ddf.spatial.clavin.index.properties"
+                               update-strategy="container-managed"/>
+        <property name="clavinIndexLocation" value="data/clavin-index"/>
+    </bean>
+
+    <service ref="clavinMetacardEnricher" interface="ddf.catalog.transform.MetacardEnricher">
+        <service-properties>
+            <entry key="id" value="clavinMetacardEnricher"/>
+            <!-- shortname only exists for backwards compatibility -->
+            <entry key="shortname" value="clavin"/>
```
Comment (millerw8) [inline catalog/spatial/geocoding/spatial-clavin-lucene/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: make id and shortname match... shortname is backwards compatible for things that look by shortname instead of id.  However, we would want those things (which may not even exist anyway) to find the same things/names regardless how they search for the services

File: catalog/spatial/geocoding/spatial-clavin-lucene/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 26]
```diff
@@ -0,0 +1,49 @@
...
+            <entry key="description"
+                   value="Enriches metacard using clavin."/>
```
Comment (millerw8) [inline catalog/spatial/geocoding/spatial-clavin-lucene/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: We can do better than this... the reader may not know what clavin is.  describe how the metacard is enriched

File: catalog/transformer/catalog-transformer-microsoft/src/main/java/ddf/catalog/transformer/input/MicrosoftInputTransformer.java [hunk 27]
```diff
@@ -0,0 +1,74 @@
...
+public class MicrosoftInputTransformer implements InputTransformer {
+    private static final Logger LOGGER = LoggerFactory.getLogger(MicrosoftInputTransformer.class);
+
+    private static int fileThreshold = 1000000;
+
+    private MetacardEnricher clavinMetacardEnricher;
+
+    private InputTransformer tikaInputTransformer;
+
+    public void setClavinMetacardEnricher(MetacardEnricher clavinMetacardEnricher) {
```
Comment (millerw8) [inline catalog/transformer/catalog-transformer-microsoft/src/main/java/ddf/catalog/transformer/input/MicrosoftInputTransformer.java]: This stuff should go away. this transformer should know nothing about Clavin.  It should just allow blueprint to inject a list of metacardEnrichers, and then call enrich on each instance.

File: catalog/core/catalog-core-api/src/main/java/ddf/catalog/transform/MetacardEnricher.java [hunk 28]
```diff
@@ -0,0 +1,41 @@
+/**
```
Comment (millerw8) [inline catalog/core/catalog-core-api/src/main/java/ddf/catalog/transform/MetacardEnricher.java]: Yeah, this is basically what I was looking for. I am not sure this is the right package to put it in or the right name to use, but can live with it for now.  I am guessing others would have opinions.  I tried to think if this concept could align with a postIngestPlugin but I don't think that is the right answer, I prefer this.

May want to put something in here that mentioned this is beta (preliminary? experimental?  what's the right word?) and subject to change

File: catalog/core/catalog-core-api/src/main/java/ddf/catalog/transform/MetacardEnricher.java [hunk 29]
```diff
@@ -0,0 +1,41 @@
...
+ * metadata to a {@link Metacard}.
+ * <p>
+ * <b>Implementations of this interface <em>must</em>:</b>
+ * <ul>
+ * <li/>Register with the OSGi Service Registry using the {@link MetacardEnricher} interface
+ * <li/>Include a Service property with name "id" ({@link ddf.catalog.Constants#SERVICE_ID}) and a
+ * {@code String} value uniquely identifying the particular implementation
+ * </ul>
+ */
+public interface MetacardEnricher {
```
Comment (stustison) [inline catalog/core/catalog-core-api/src/main/java/ddf/catalog/transform/MetacardEnricher.java]: I don't think that we actually need this interface. If we use the CatalogFramework.create(CreateStorageRequest) method, that will allow you to ingest a microsoft document, generate a metacard from it using your transformer and then as the document itself is getting added to the content store, you can parse it in a storage plugin to add additional fields to the metacard. The PostCreateStoragePlugins hand you a list of content items and each content item has a metacard and an InputStream, which is exactly what this interface uses.

File: catalog/spatial/geocoding/spatial-clavin-lucene/src/main/java/org/codice/ddf/spatial/clavin/ClavinMetacardEnricher.java [hunk 30]
```diff
@@ -0,0 +1,84 @@
...
+
+    @Override
+    public void enrich(Metacard metacard, InputStream input) throws CatalogTransformerException {
+        try {
+            File directoryFile = new File(clavinIndexLocation);
+            LOGGER.info("clavin index: " + directoryFile.getAbsolutePath());
+            GeoParser clavinParser = GeoParserFactory.getDefault(directoryFile.getAbsolutePath());
+
+            // TODO: chunk inputstream into strings.
+            String document = IOUtils.toString(input, "UTF-8");
```
Comment (stustison) [inline catalog/spatial/geocoding/spatial-clavin-lucene/src/main/java/org/codice/ddf/spatial/clavin/ClavinMetacardEnricher.java]: You can't do this, this will blow up the JVM on a sufficiently large document.

Comment (glenhein) [inline catalog/spatial/geocoding/spatial-clavin-lucene/src/main/java/org/codice/ddf/spatial/clavin/ClavinMetacardEnricher.java]: When you do chunk the data, be careful since UTF8 is multibyte, so you can't chunk on simple byte boundaries.

File: catalog/core/catalog-core-api/src/main/java/ddf/catalog/transform/MetacardEnricher.java (hunk 29 above)
Comment (millerw8) [inline catalog/core/catalog-core-api/src/main/java/ddf/catalog/transform/MetacardEnricher.java]: cool, so what we were trying to do already exists, and it doesn't have to be processed by the transformer.  That's great,  just no one knew about it.  Where were you last week :stuck_out_tongue: ?

**Comment by shane-voisard** [comment]
Close this pull request because the design changed from a transformer to a post ingest plugin.
